include/fsm/fsm.h

```diff
@@ -201,6 +206,41 @@ fsm_setopaque(struct fsm *fsm, fsm_state_t state, void *opaque);
 void *
 fsm_getopaque(const struct fsm *fsm, fsm_state_t state);
 
+/* Associate a numeric ID with the end states in an fsm.
+ * This can be used to track which of the original fsms matched
+ * input when multiple fsms are combined.
+ *
+ * These will be preserved through the following operations:
+ * - determinise
+ * - union
+ * - concat
+ * - ...
+ * */
+int
+fsm_setendid(struct fsm *fsm, fsm_end_id_t id);
```

Comment:
What happens for other operations? Do they get cleared?

Reply:
I'm going to refine that comment in the next pass. If I can't support something I would rather clear it and explicitly note it there. We could also clear the capture actions if we detect that we're minimising after a union.